### PR TITLE
Add context menu option to list programs in a series

### PR DESCRIPTION
**Pull request by mlehtima**
Requires https://github.com/joonne/harbour-nayttamo/pull/24

### CODE REVIEW

File: qml/js/yleApi.js [hunk 1]
```diff
@@ -108,7 +108,7 @@ var api = (function(appId, appKey) {
                                ? program.partOfSeason.seasonNumber
                                : "",
                  episodeNumber: program.episodeNumber ? program.episodeNumber : "",
-                 seriesId: program.partOfSeason && program.partOfSeason.id,
+                 seriesId: program.partOfSeries && program.partOfSeries.id
```
Comment (joonne) [inline qml/js/yleApi.js:111]: so `partOfSeason` is then something that we can use later on and find programs by season?

File: qml/pages/ProgramsPage.qml [hunk 2]
```diff
@@ -7,62 +7,69 @@ Page {
...
-                console.log('programs', programs)
-                listView.model = programs
-            })
-            .catch(function(error) {
-                console.log('error', error)
-                listView.model = []
-            })
+        if (category.id) {
+            YleApi.getProgramsByCategoryId(category.id, limit, offset)
+                .then(function(programs) {
```
Comment (joonne) [inline qml/pages/ProgramsPage.qml]: extract this function since it is duplicated here and in the alternative branch?

File: qml/js/yleApi.js (hunk 1 above)
Comment (mlehtima) [reply qml/js/yleApi.js:111]: We already use `program.partOfSeason.seasonNumber `a few lines above, this was probably a bug.

File: qml/pages/ProgramsPage.qml (hunk 2 above)
Comment (mlehtima) [reply qml/pages/ProgramsPage.qml]: Good point, I'll fix that

File: qml/js/yleApi.js (hunk 1 above)
Comment (joonne) [reply qml/js/yleApi.js:111]: :+1:

File: qml/pages/ProgramsPage.qml (hunk 2 above)
Comment (joonne) [reply qml/pages/ProgramsPage.qml]: :+1:

**Comment by mlehtima** [comment]
I made the search field get automatic focus so the virtual keyboard would appear automatically, wondering if we want to have that.

**Comment by joonne** [comment]
We can try that, I kinda think that the self appearing keyboard is annoying and the placeholder text "invites" the user to do the search. No strong opinions though :)

File: qml/js/yleApi.js [hunk 3]
```diff
@@ -126,6 +126,14 @@ var api = (function(appId, appKey) {
              .then(mapPrograms);
      }
 
+     function getProgramsBySeriesId(seriesId, limit, offset) {
+         var url = apiUrl + programsUrl + '&series=' + seriesId + '&availability=ondemand&mediaobject=video&offset=' + offset + '&limit=' + limit;
```
Comment (joonne) [inline qml/js/yleApi.js]: do we need the contentprotection for this also: `&contentprotection=22-0,22-1&availability=ondemand` ?

Comment (mlehtima) [reply qml/js/yleApi.js]: Yes, it should be there, I will move those common parameters to the `programsUrl` variable.

**Comment by mlehtima** [comment]
I removed the focus change because it caused some overlap with keyboard. I also added a check to make sure publication event has media available to minimize risk of non-working videos.

**Comment by mlehtima** [comment]
I added also series listing context menu to first page of the app for current broadcasts. Would you have time to review this so we could make a new release early next week? The current version in store is some much worse.

**Comment by joonne** [comment]
I'll review and test this during next week for sure, I have a deadline for one other hobby project for Wednesday but it should be ready today/tomorrow. Then I should have more time for näyttämö :)

File: qml/js/yleApi.js [hunk 4]
```diff
@@ -8,7 +8,7 @@ var api = (function(appId, appKey) {
      // .env -> qmake -> DEFINES -> setContextProperty
      var credentials = "app_id=" + appId + "&app_key=" + appKey;
 
-     var programsUrl = "/programs/items.json?" + credentials;
+     var programsUrl = "/programs/items.json?" + credentials + "&contentprotection=22-0,22-1&availability=ondemand&mediaobject=video";
```
Comment (joonne) [inline qml/js/yleApi.js]: :+1:

Comment (joonne) [inline qml/js/yleApi.js:11]: :+1:

File: qml/js/yleApi.js [hunk 5]
```diff
@@ -26,7 +26,7 @@ var api = (function(appId, appKey) {
      var filterAvailablePrograms = function(programs) {
          return programs.reduce(function(acc, program) {
              var foundPublicationEvents = program.publicationEvent.filter(function(event) {
-                 return event.temporalStatus === "currently" && event.type === "OnDemandPublication";
+                 return event.temporalStatus === "currently" && event.type === "OnDemandPublication" && event.media && event.media.available;
```
Comment (joonne) [inline qml/js/yleApi.js:29]: This was missing also? :thinking:

File: qml/pages/FirstPage.qml [hunk 6]
```diff
@@ -62,6 +62,14 @@ Page {
                                            })
                         }
                     }
+                    MenuItem {
+                        visible: modelData.seriesId ? true : false
```
Comment (joonne) [inline qml/pages/FirstPage.qml]: useless ternary but could be replaced with `Boolean(modelData.seriesId)`?

**Review by joonne — APPROVED** [review]
Looks good! The new functionality with showing programs in series is amazing, good job! :)

**Comment by joonne** [comment]
As a general note we could think of adding a spinner or some other indication when the autoloading new items, I'll add a ticket for it.

File: qml/js/yleApi.js (hunk 5 above)
Comment (mlehtima) [reply qml/js/yleApi.js:29]: I saw some videos being listed without a event.media value and some without available media.

File: qml/pages/FirstPage.qml (hunk 6 above)
Comment (mlehtima) [reply qml/pages/FirstPage.qml]: Changed this and another similar line to use Boolean function.

**Comment by mlehtima** [comment]
A spinner would probably be a good idea. I pushed the boolean changes and also removed one unused API call from ProgramOverviewPage.

File: qml/pages/ProgramOverviewPage.qml [hunk 7]
```diff
@@ -8,10 +8,6 @@ Page {
 
     property var program: ({})
 
-    Component.onCompleted: {
```
Comment (joonne) [inline qml/pages/ProgramOverviewPage.qml:11]: How was this even there anymore :sweat_smile:

**Comment by joonne** [comment]
looks good, ready to merge :)

**Comment by mlehtima** [comment]
I think now would be a good time to make a new release. Could you do it when you have time. Check also the issues about supported Sailfish OS versions in store and also x86 build would be nice. I just tested the app on Jolla Tablet and it worked well. If you want me to test the x86 build you make send the rpm via email.

**Pull request merged**